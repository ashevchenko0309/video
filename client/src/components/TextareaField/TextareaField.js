import React, { useState, useCallback } from "react"
import PropTypes from "prop-types"
import { throttle } from "lodash"
import classNames from "classnames"

import FormFieldErrorHint from "../FormErrorHints/FormFieldErrorHint"

function TextareaField({
  name,
  defaultValue,
  placeholder,
  labelText,
  isRequired,
  options,
  hasError,
  delay,
  errorMessage,
  onChange,
  onBlur,
}) {
  const [textareaValue, setTextareaValue] = useState(defaultValue)

  const throttledTextareaChange = useCallback(
    throttle((v) => onChange(v), delay),
    []
  )

  const defaultValidation = { minLength: isRequired ? 1 : 0, maxLength: 255 }

  const validationOptions = options || defaultValidation

  const formFieldClass = classNames({
    form_field: true,
    "has-error": hasError,
  })

  const onChangeHandler = (e) => {
    const { value } = e.currentTarget
    setTextareaValue(value)
    throttledTextareaChange(value)
  }

  const onBlurHandler = (e) => {
    const { value } = e.currentTarget
    onBlur(value, name)
  }

  return (
    <div className={formFieldClass}>
      <label htmlFor={name}>
        <span className="input_field--label">
          {labelText}
          {isRequired && <span className="is-required"> *</span>}
        </span>
        <span className="input_field--hint">
          {textareaValue.length} / {validationOptions.maxLength}
        </span>
        <textarea
          className="form_field--textarea"
          value={textareaValue}
          name={name}
          placeholder={placeholder}
          required={isRequired}
          maxLength={validationOptions.maxLength}
          onChange={onChangeHandler}
          onBlur={onBlurHandler}
        />
      </label>
      <FormFieldErrorHint hasError={hasError} errorMessage={errorMessage} />
    </div>
  )
}

TextareaField.propTypes = {
  name: PropTypes.string,
  defaultValue: PropTypes.string,
  placeholder: PropTypes.string,
  labelText: PropTypes.string,
  errorMessage: PropTypes.string,
  isRequired: PropTypes.bool,
  hasError: PropTypes.bool,
  delay: PropTypes.number,
  options: PropTypes.shape({}),
  onChange: PropTypes.func,
  onBlur: PropTypes.func,
}

TextareaField.defaultProps = {
  name: "",
  defaultValue: "",
  placeholder: "",
  labelText: "",
  isRequired: false,
  options: null,
  hasError: false,
  delay: 150,
  errorMessage: "",
  onChange: () => {},
  onBlur: () => {},
}

export default TextareaField
