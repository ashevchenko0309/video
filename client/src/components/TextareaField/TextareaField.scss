.form_field--textarea{
  display: block;
  width: 100%;
}