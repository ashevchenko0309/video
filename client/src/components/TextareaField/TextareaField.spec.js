import React from "react"
import TextareaField from "./TextareaField"

const DEFAULT_PROPS = {
  name: "Test",
  placeholder: "Test placeholder",
  labelText: "Test label",
}

const setUp = (props) => mount(<TextareaField {...props} />)

describe("TextareaField component", () => {
  let component

  beforeEach(() => {
    component = setUp({ ...DEFAULT_PROPS })
  })

  it("should render default textarea (snapshot)", () => {
    expect(component.getElement()).toMatchSnapshot()
  })

  it("should render textarea with default value (snapshot)", () => {
    const component = setUp({ ...DEFAULT_PROPS, defaultValue: "default value" })
    expect(component.getElement()).toMatchSnapshot()
  })

  it("should render textarea with error message (snapshot)", () => {
    const component = setUp({
      ...DEFAULT_PROPS,
      hasError: true,
      errorMessage: "Some error",
    })
    expect(component.getElement()).toMatchSnapshot()
  })

  it("should render textarea with all props (snapshot)", () => {
    const component = setUp({
      ...DEFAULT_PROPS,
      defaultValue: "default value",
      hasError: true,
      errorMessage: "Some error",
    })
    expect(component.getElement()).toMatchSnapshot()
  })

  it("should have default length hint", () => {
    expect(component.find(".input_field--hint").text()).toBe("0 / 255")
  })

  it("should have required hint", () => {
    const textareaField = setUp({ ...DEFAULT_PROPS, isRequired: true })
    expect(textareaField.find(".input_field--label").text()).toBe(
      "Test label *"
    )
  })

  it("should have default inactive class error hint", () => {
    const textareaField = setUp({ ...DEFAULT_PROPS, isRequired: true })
    expect(textareaField.find(".input_field--error").hasClass("inactive")).toBe(
      true
    )
  })

  it("should have active class error hint", () => {
    const textareaField = setUp({
      ...DEFAULT_PROPS,
      isRequired: true,
      hasError: true,
      errorMessage: "Test error",
    })
    expect(textareaField.find(".form_field").hasClass("has-error")).toBe(true)
    expect(textareaField.find(".input_field--error").hasClass("active")).toBe(
      true
    )
  })

  it("should have error hint", () => {
    const textareaField = setUp({
      ...DEFAULT_PROPS,
      isRequired: true,
      hasError: true,
      errorMessage: "Test error",
    })
    expect(textareaField.find(".input_field--error").text()).toBe("Test error")
  })

  it("shold call setInputValue method", () => {
    const setTextareaValue = jest.fn()
    const inputField = setUp({ ...DEFAULT_PROPS, onChange: setTextareaValue })
    const handleType = jest.spyOn(React, "useState")
    handleType.mockImplementation((textareaValue) => [
      textareaValue,
      setTextareaValue,
    ])

    expect(setTextareaValue.mock.calls.length).toBe(0)

    inputField
      .find(".form_field--textarea")
      .simulate("change", { currentTarget: { value: "123456789" } })
    expect(setTextareaValue).toBeTruthy()
    expect(setTextareaValue.mock.calls.length).toBe(1)
  })

  it("should call onBlur method", () => {
    const onBlur = jest.fn()
    const inputField = setUp({ ...DEFAULT_PROPS, onBlur })

    expect(onBlur.mock.calls.length).toBe(0)
    inputField
      .find(".form_field--textarea")
      .simulate("blur", { currentTarget: { value: "123456789" } })
    expect(onBlur.mock.calls.length).toBe(1)
  })

  it("should use default onChange prop", () => {
    const result = TextareaField.defaultProps.onChange()
    expect(result).toBe(undefined)
  })

  it("should use default onBlur prop", () => {
    const result = TextareaField.defaultProps.onBlur()
    expect(result).toBe(undefined)
  })
})
