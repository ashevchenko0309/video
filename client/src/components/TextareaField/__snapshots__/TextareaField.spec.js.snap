// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextareaField component should render default textarea (snapshot) 1`] = `
<TextareaField
  defaultValue=""
  delay={150}
  errorMessage=""
  hasError={false}
  isRequired={false}
  labelText="Test label"
  name="Test"
  onBlur={[Function]}
  onChange={[Function]}
  options={null}
  placeholder="Test placeholder"
/>
`;

exports[`TextareaField component should render textarea with all props (snapshot) 1`] = `
<TextareaField
  defaultValue="default value"
  delay={150}
  errorMessage="Some error"
  hasError={true}
  isRequired={false}
  labelText="Test label"
  name="Test"
  onBlur={[Function]}
  onChange={[Function]}
  options={null}
  placeholder="Test placeholder"
/>
`;

exports[`TextareaField component should render textarea with default value (snapshot) 1`] = `
<TextareaField
  defaultValue="default value"
  delay={150}
  errorMessage=""
  hasError={false}
  isRequired={false}
  labelText="Test label"
  name="Test"
  onBlur={[Function]}
  onChange={[Function]}
  options={null}
  placeholder="Test placeholder"
/>
`;

exports[`TextareaField component should render textarea with error message (snapshot) 1`] = `
<TextareaField
  defaultValue=""
  delay={150}
  errorMessage="Some error"
  hasError={true}
  isRequired={false}
  labelText="Test label"
  name="Test"
  onBlur={[Function]}
  onChange={[Function]}
  options={null}
  placeholder="Test placeholder"
/>
`;
